### ISSUE TITLE
General support for out-of-tree shared task libraries

### ISSUE THREAD

**Issue by sm-Fifteen** [PrefectHQ/prefect#4175]
## Current behavior
The Prefect task library currently contains task wrappers for a lot of common Python libraries ([some of then trvial](https://github.com/PrefectHQ/prefect/blob/2b6fee35a595170c06fa813a9366192b5de64be3/src/prefect/tasks/rss/feed.py)) in order to let users focus on writing Flows and keep the amount of user-defined tasks in the flow to only the things that likely can't be reused. This is something [the documentation seems to encourage](https://docs.prefect.io/core/task_library/contributing.html), saying that users should by all means try to upstream any useful tasks that the library doesn't cover yet. Besides creating [a lot of optional dependencies](https://github.com/PrefectHQ/prefect/blob/2b6fee35a595170c06fa813a9366192b5de64be3/setup.py#L41-L83), however, one of the problems of having the in-tree task library be the repository for all shared tasks by default means that there is currently no mechanism within Prefect to let users define alternate or private task libraries, and [dependency management for user-defined shared tasks can become complicated as a result](https://github.com/PrefectHQ/prefect/discussions/4172).

## Proposed behavior
Prefect should have a way of letting users create and manage their own collections of tasks to be reused across agents. Such a mechanism could also be used to reduce the core task library to a more basic subset and move the community-contributed or third-party ones elsewhere.

The way I envision it, task modules should be able to be registered on a flow to ensure they are fetched and available by the time the flow runs.

```py
oracle_tasks = LocalTaskModule('/opt/prefect/shared_tasks/oracle') # Must already be present on the agent
pandas_tasks = BitbucketTaskModule(path="tasks/pandas", repo="prefect-flows", ref='bdef1234') # Will be fetched by the agent
dropbox_tasks = CommunityTaskModule('dropbox') # Fetched from some official community task repository??

with Flow('My Flow'):
    dropbox_tasks.DropboxDownload(...)
```

Basically, some way to have the task dependencies be known and possibly managed by the agent or RunConfig instead of having either `Docker` storage take care of everything at build time or use another storage method and have to manually configure your agents (using `--import-path` that feature is not supported in the prefect config) for all tasks it's expected to be able to run (and having labels to identify which tasks are registered on each agent, etc.).

Ideally, agents would be able to automatically setup requirements (using pip, for instance) for those task modules as well, be it in virtual environments or in docker images or whatever the optimal way of handling this is depending on how flows are expected to be run.

**Comment by sm-Fifteen** [comment]
This seems to be addressed in Prefect 2.0 by the [new](https://www.prefect.io/blog/collections-in-prefect-2-0) (and [confusingly named](https://docs.prefect.io/core/concepts/tasks.html#collections)) [Prefect collections](https://orion-docs.prefect.io/collections/overview/). It's still unclear to me how these are meant to interact with agent deployments and how this affects "private" (i.e: not uploaded on pip) collections, but that should be the topic of a different issue.

**Comment by madkinsz** [comment]
Hey @sm-Fifteen :) curious why you find the name confusing? I can probably clarify some of the reasoning and future intent for you and we can update the documentation accordingly.

**Comment by sm-Fifteen** [comment]
@madkinsz: Oh, nothing much, it's just that, since collection tasks were already a thing in Prefect v1 (not sure if they still exist in Prefect 2?), looking up "prefect collections" is likely to give you both results.

As for the questions I had (I was about to open a discussion thread on the matter), I'm interested to know how dependency management for these works with Prefect Agents. Right now, with Prefect v1, you just need manually install all dependencies your flows may require on all runners that should have it (unless you're running docker images), and possibly add tags representing these dependencies on the Prefect runner (which requires restarting it) and the tasks so flows aren't sent to runners that can't run them. I'm curious if Prefect Collections currently do anything to make dependency management better automated, or if there's any plans of such a feature.

I'm also interested in having shared private libraries of utility tasks, and the doc and templates really seem to put the emphasis on pip, so I would like to know if that's something that's currently possible with v2 or in the plans.

**Comment by madkinsz** [comment]
Ah, yeah we were a bit concerned about that collision but the necessity of "collection" tasks for interacting with Python collections will dissipate in v2 since you can just work with collections dynamically in a flow with native Python. We wanted a name that would capture the inclusion of more than just tasks as we intend to expose a myriad of integrations in each collection library.

V2 supports virtual and conda environment specification per flow run which should help some. We're working on more robust workflow packaging in the near future so this should be addressed more comprehensively in the future. I don't think collections will help with more automated dependency management since they're just Python packages but we are interested in automatic dependency management for Python packages in general.

We're hoping to address private libraries of shared tasks with better Git repository integrations as well as the packaging milestone mentioned above.

Thanks for your feedback!

**Comment by sm-Fifteen** [comment]
> V2 supports virtual and conda environment specification per flow run which should help some. We're working on more robust workflow packaging in the near future so this should be addressed more comprehensively in the future. I don't think collections will help with more automated dependency management since they're just Python packages [...].

Yeah, I understand the sort of complexities (not to mention security concerns?) that would come with some kind of "auto-pip" behavior. I figured Agents automatically adding tags based on the discovered (if that's even doable? I figure there might be options if all the collections follow some established template) Prefect collections that are installed in the Agent's virtual environment, and importing collections during flow registration implicitly adding the same tags to that flow, somehow.

Probably more trouble than it's worth, but it certainly has the potential be useful while waiting for flow packaging to be finalized.

> We're hoping to address private libraries of shared tasks with better Git repository integrations as well as the packaging milestone mentioned above.

I'm eager to see the results. Docker images on runners aren't an option on my current deployment setup, so I'm currently falling back on having all flows reside in a single module at the root of the agent's working directory, so that flows can use absolute imports along the lines of `from everything_prefect.shared_utils.database import SybaseQuery` or something. It works, but it's a lot more brittle than I'd like and requires a lot of manual intervention on the Prefect runners (of which I thankfully only have one).
